### PR TITLE
Implement LWG-3710: The `end` of `chunk_view` for input ranges can be `const`

### PR DESCRIPTION
**Pull request by GoGim1**
Implements the proposed resolution of [LWG-3710](https://cplusplus.github.io/LWG/issue3710)

### CODE REVIEW

**Comment by StephanTLavavej** [comment]
Welcome to the microsoft/STL repo!

The `P2442R1_views_chunk` test is failing with:

```
C:\a\1\s\tests\std\tests\P2442R1_views_chunk\test.cpp(200): error C2338: static_assert failed: 'CanMemberEnd<const R> == forward_range<const V>'
C:\a\1\s\tests\std\tests\P2442R1_views_chunk\test.cpp(204): error C2338: static_assert failed: 'common_range<R> == (common_range<const V> && (sized_range<const V> || !bidirectional_range<const V>) )'
C:\a\1\s\tests\std\tests\P2442R1_views_chunk\test.cpp(202): error C7601: the associated constraints are not satisfied
```

Please ensure that this test passes locally before pushing changes. Let us know if you need assistance!

**Comment by frederick-vs-ja** [comment]
I think this test file should also be changed.

https://github.com/microsoft/STL/blob/ef62d3fa0b8e4e2406b9bb74e916e1ca8a1df802/tests/std/tests/P2442R1_views_chunk/test.cpp#L200

Due to LWG-3710, `CanMemberEnd<const R>` becomes `true` if `const V` is an `input_range`.

File: tests/std/tests/P2442R1_views_chunk/test.cpp [hunk 1]
```diff
@@ -197,12 +197,13 @@ constexpr bool test_one(Rng&& rng, Expected&& expected) {
...
 
-    STATIC_ASSERT(CanMemberEnd<const R> == forward_range<const V>);
+    STATIC_ASSERT(CanMemberEnd<const R> == input_range<const V>);
     if constexpr (CanMemberEnd<const R>) {
         const same_as<sentinel_t<const R>> auto cs = as_const(r).end();
         assert((r.begin() == cs) == is_empty);
         STATIC_ASSERT(
-            common_range<R> == (common_range<const V> && (sized_range<const V> || !bidirectional_range<const V>) ));
+            common_range<
+                R> == (forward_range<const V> && common_range<const V> && (sized_range<const V> || !bidirectional_range<const V>) ));
```
Comment (GoGim1) [inline tests/std/tests/P2442R1_views_chunk/test.cpp:206]: I ran into a problem that clang-format wants this line to look like this,, while Code Format Validation will complain the line more than 120 columns. Any ideas?

Comment (GoGim1) [reply tests/std/tests/P2442R1_views_chunk/test.cpp:206]: The following are error messages from workflows:
[Expected formatting](https://dev.azure.com/vclibs/STL/_build/results?buildId=11717&view=logs&j=ce9c9df7-f6fa-59cf-0934-3339753bfc56&t=c6e68a6e-a0d3-5ff7-0941-3a725df1e075&l=20)
[Code Format Validation failed](https://dev.azure.com/vclibs/STL/_build/results?buildId=11719&view=logs&j=ce9c9df7-f6fa-59cf-0934-3339753bfc56&t=587b9e2b-1c42-5f9c-5a6a-63448075a96c&l=13)

Comment (miscco) [inline tests/std/tests/P2442R1_views_chunk/test.cpp:206]: You could try to force clang-formats hand by adding a comment
```suggestion
            common_range<R> //
                == (forward_range<const V> && common_range<const V> && (sized_range<const V> || !bidirectional_range<const V>) ));
```

Comment (StephanTLavavej) [reply tests/std/tests/P2442R1_views_chunk/test.cpp:206]: No problem, this is definitely an annoying issue. I second @miscco's recommendation https://github.com/microsoft/STL/pull/2878#discussion_r924459742 - adding an empty comment will force clang-format to wrap, and is typically the least intrusive way to fix this.

#2475 is our tracking issue for clang-format headaches like this, and they will reportedly get better in the next release of clang-format.

**Comment by CaseyCarter** [comment]
This looks good, other than the formatting issue.

File: tests/std/tests/P2442R1_views_chunk/test.cpp [hunk 2]
```diff
@@ -197,12 +197,14 @@ constexpr bool test_one(Rng&& rng, Expected&& expected) {
         }
     }
 
-    STATIC_ASSERT(CanMemberEnd<const R> == forward_range<const V>);
+    STATIC_ASSERT(CanMemberEnd<const R> == input_range<const V>);
     if constexpr (CanMemberEnd<const R>) {
         const same_as<sentinel_t<const R>> auto cs = as_const(r).end();
```
Comment (GoGim1) [inline tests/std/tests/P2442R1_views_chunk/test.cpp:202]: Sorry to bother you again. I wonder why the associated constraints are not satisfied in this line. It seems that nothing wrong with the unit test. Maybe it due to my incorrect implementation? @StephanTLavavej

Comment (StephanTLavavej) [reply tests/std/tests/P2442R1_views_chunk/test.cpp:202]: I'm not immediately sure what's happening here, but @CaseyCarter will look into this and reply soon. :smiley_cat:

Comment (CaseyCarter) [reply tests/std/tests/P2442R1_views_chunk/test.cpp]: `sentinel_t<T>` is constrained to require `range<T>`. Previously, `CanMemberBegin<T> == CanMemberEnd<T>` was always `true`, so `CanMemberEnd<const R>` implied `range<const R>`, but this is no longer the case. It's now possible that `CanMemberEnd<const R> && !CanMemberBegin<const R>`, in which case `sentinel_t` is ill-formed and this line catches fire.

The fix (which I've applied) is to nest this block inside `if constexpr (CanMemberBegin<const R>)`, and add an `else` branch that validates only that `end()` returns `default_sentinel_t`.

File: tests/std/tests/P2442R1_views_chunk/test.cpp (hunk 1 above)
Comment (CaseyCarter) [reply tests/std/tests/P2442R1_views_chunk/test.cpp]: I tried some other options to make this less ugly, but finally gave up and dropped the `// clang-format off` hammer.

**Review by CaseyCarter — APPROVED** [review]

File: tests/std/tests/P2442R1_views_chunk/test.cpp [hunk 3]
```diff
@@ -197,23 +197,30 @@ constexpr bool test_one(Rng&& rng, Expected&& expected) {
...
-                assert(equal(*prev(cs), *prev(end(expected))));
-            }
-
-            if constexpr (copy_constructible<V>) {
-                const auto r2 = r;
+        if constexpr (CanMemberBegin<const R>) {
+            const same_as<sentinel_t<const R>> auto cs = as_const(r).end();
+            assert((r.begin() == cs) == is_empty);
+            // clang-format off
+            STATIC_ASSERT(common_range<const R>
```
Comment (CaseyCarter) [inline tests/std/tests/P2442R1_views_chunk/test.cpp:206]: Note the drive-by fix for the pre-existing bug: this was not validating `common_range<const R>`, but checking that `common_range<R>` is true when the conditions necessary for `common_range<const R>` hold. (@miscco, please check my math.) (No change requested.)

Comment (miscco) [reply tests/std/tests/P2442R1_views_chunk/test.cpp:206]: Yeah, it seems that one slipped through. Thanks for fixing

**Review by miscco — APPROVED** [review]

**Comment by GoGim1** [comment]
Thanks everyone. Without your help, I could not have finished this patch.

File: tests/std/tests/P2442R1_views_chunk/test.cpp [hunk 4]
```diff
@@ -197,23 +197,30 @@ constexpr bool test_one(Rng&& rng, Expected&& expected) {
...
+                == (forward_range<const V> && common_range<const V> &&
```
Comment (StephanTLavavej) [inline tests/std/tests/P2442R1_views_chunk/test.cpp:207]: Nitpick: We conventionally put binary operators at the beginning of a line, where they are more easily seen. (clang-format would do this automatically, but it's suppressed here.)

It should be possible to deal with suboptimal clang-formatting here without entirely suppressing it. Adding empty comments like `//` will force line breaks.

File: tests/std/tests/P2442R1_views_chunk/test.cpp [hunk 5]
```diff
@@ -197,23 +197,30 @@ constexpr bool test_one(Rng&& rng, Expected&& expected) {
...
+
+                if constexpr (copy_constructible<V>) {
+                    const auto r2 = r;
+                    if (!is_empty) {
+                        assert(equal(*prev(r2.end()), *prev(end(expected))));
+                    }
                 }
             }
+        } else {
+            STATIC_ASSERT(same_as<decltype(as_const(r).end()), std::default_sentinel_t>);
```
Comment (StephanTLavavej) [inline tests/std/tests/P2442R1_views_chunk/test.cpp:223]: Nitpick: This test is already `using namespace std;` on line 14.

**Review by StephanTLavavej — COMMENTED** [review]
Thanks, this looks great! I'll push minor changes for nitpicky style issues.

**Review by StephanTLavavej — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
FYI @CaseyCarter, I pushed very minor changes after you approved.

**Review by CaseyCarter — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
I'm mirroring this to the MSVC-internal repo - please notify me if any further changes are pushed.

**Comment by StephanTLavavej** [comment]
Thanks for improving `<ranges>` conformance, and congratulations on your first microsoft/STL commit! :heart_eyes_cat: :tada: :rocket:

This change will be available in VS 2022 17.4 Preview 2.
